docs/src/developer_documentation/ir_representation.md

```diff
@@ -0,0 +1,560 @@
+
+julia> insert!(bb_ir_copy.blocks[3], 1, new_id, ni)
+
+julia> bb_ir_copy.blocks[3].insts[2].stmt.args[2] = new_id
+ID(108)
+
+julia> CC.IRCode(bb_ir_copy)
+  1 ─      nothing::Nothing
+4 2 ┄ %2 = φ (#1 => 1, #3 => %8)::Int64
+  │   %3 = φ (#1 => 0, #3 => %7)::Int64
+  │   %4 = Base.slt_int(%3, _2)::Bool
+  └──      goto #4 if not %4
+2 3 ─ %6 = (Core.Intrinsics.mul_int)(%3, 2)::Int64
+5 │   %7 = Base.add_int(%6, 1)::Int64
+6 │   %8 = (Core.Intrinsics.add_int)(%2, %7)::Int64
+7 └──      goto #2
+8 4 ─      return %2
+```
+We see here that `IRCode` and `BBCode` involve similar levels of complexity to insert an instruction.
+
+### Inserting New Basic Blocks
+
+This is the situation in which the design of `BBCode` shines vs `IRCode`.
+`IRCode` does not, at present, really have much to say about transformations which change control flow.
+It is, however, straightforward using `BBCode`.
+Suppose that we wish to modify the above to display the value of `%2` if it is even on any given iteration.
+Since this involves control flow, it necessarily requires at least one additional basic block.
+
+We do this in two steps.
+We first insert an additional basic block between blocks `#3` and `#4` which always prints out the value of `%2`, and then goes to block `#2`:
+```jldoctest my_factorial
+julia> using Mooncake.BasicBlockCode: BBlock, new_inst, IDGotoNode, IDGotoIfNot
+
+julia> block_2_id = bb_ir_copy.blocks[2].id;
+
+julia> new_bb_id = ID();
+
+julia> new_bb = BBlock(
+           new_bb_id,
+           ID[ID(), ID()],
+           CC.NewInstruction[
+               new_inst(Expr(:call, println, CC.SSAValue(2))),
+               new_inst(IDGotoNode(block_2_id)),
+           ],
+       );
+
+julia> insert!(bb_ir_copy.blocks, 4, new_bb);
+
+julia> CC.IRCode(bb_ir_copy)
+  1 ─      nothing::Nothing
+4 2 ┄ %2 = φ (#1 => 1, #3 => %8)::Int64
+  │   %3 = φ (#1 => 0, #3 => %7)::Int64
+  │   %4 = Base.slt_int(%3, _2)::Bool
+  └──      goto #5 if not %4
+2 3 ─ %6 = (Core.Intrinsics.mul_int)(%3, 2)::Int64
+5 │   %7 = Base.add_int(%6, 1)::Int64
+6 │   %8 = (Core.Intrinsics.add_int)(%2, %7)::Int64
+7 └──      goto #2
+2 4 ─      (println)(%2)::Any
+  └──      goto #2
+8 5 ─      return %2
+```
+Observe that, in this case, rather than creating `new_bb` and then inserting instructions into it, we simply create the block _with_ the instructions.
+This programming style is often convenient.
+Additionally note that we create an `ID` for each statement in the new basic block.
+These `ID`s are never actually used anywhere, but `BBCode` requires that each instruction be associated to an `ID`, so we must create them.
+
+Additionally, note the usage of an `IDGotoNode`.
+This is exactly the same thing as a `Core.Compiler.GotoNode`, except it contains an `ID` stating which basic block to jump to, rather than an `Int`.
+Similarly, the `IDGotoIfNot` is a direct translation of `Core.Compiler.GotoIfNot`, with the `dest` field being an `ID` rather than an `Int`.
```

Comment:
```suggestion
Similarly, the [`IDGotoIfNot`](@ref) is a direct translation of `Core.Compiler.GotoIfNot`, with the `dest` field being an `ID` rather than an `Int`.
```